### DISCUSSION TITLE
Troubles using Triangle Shapes

### DISCUSSION THREAD

**Discussion by alecpizz** [notgiven688/jitterphysics2#208] [How do I ... ?]
Hello, I apologize if I'm in the wrong place. I'm currently playing around with Jitter, and would like to apply a general static collision mesh to my scene. I'm using raylib currently, and I build my triangle shape as follows.

```csharp
 //load in a quake map
 Model quakeMap = LoadModel(@"Resources\Models\Quake\scene.gltf");
 RigidBody citybody = world.CreateRigidBody();
 List<JTriangle> tris = new List<JTriangle>();
 for (int i = 0; i < city.MeshCount; i++)
 {
     var mesh = city.Meshes[i];
     //based on Raylibs GetRayCollisionMeshes
     Vector3* vertdata = (Vector3*)mesh.Vertices;
     if (mesh.Indices != null)
     {
         for (int j = 0; j < mesh.TriangleCount; j++)
         {
             //ToJVector is just a sys.numerics converter helper
             JVector a = vertdata[mesh.Indices[j * 3 + 0]].ToJVector();
             JVector b = vertdata[mesh.Indices[j * 3 + 1]].ToJVector();
             JVector c = vertdata[mesh.Indices[j * 3 + 2]].ToJVector();
             JVector normal = (c - a) % (b - a);
             //I was getting the occasional degenerate triangle exception here.
             if (MathHelper.CloseToZero(normal, 1e-12f))
             {
                   continue;
             }

             tris.Add(new JTriangle(a, b, c));
        }
     }
 }
var jtm = new TriangleMesh(tris);
List<RigidBodyShape> triangleShapes = new List<RigidBodyShape>();
for (int i = 0; i < jtm.Indices.Length; i++)
{
        TriangleShape ts = new TriangleShape(jtm, i);
        triangleShapes.Add(ts);
}
            
citybody.AddShape(triangleShapes, false);
citybody.Position = city.Transform.Translation.ToJVector();   
citybody.IsStatic = true;
```
I've found that my triangle generation is generally correct for something simple like a single cube, where when I spawn in a cube, it will fall within the room.
For my more complex scene, I get some strange behavior:

https://github.com/user-attachments/assets/ad3b8cdb-6543-462a-8b58-1f346d1777ec

I can't draw the scene using jitter's debug drawer as it freezes my application. 

I've enjoyed working with this library so far, and I'd like to continue using it, so any help would be great!

**Comment by notgiven688** [comment]
Looks good. I think your triangles are just facing into the wrong direction. Try:

`tris.Add(new JTriangle(b, a, c));`

**Reply by alecpizz** [reply]
![image](https://github.com/user-attachments/assets/ac416ce8-62d6-4e81-adbd-6838bcd041ff)
I knew it was something simple, that seems to have done the trick, thank you!
